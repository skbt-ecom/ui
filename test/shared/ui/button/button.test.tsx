import { render, screen } from '@testing-library/react'
import { describe, expect, it } from 'vitest'
import { Button } from '$/shared/ui'

describe('Тестирование Button', () => {
  const buttonTestid = 'test-button'

  it('Кнопка должна рендериться с правильным текстом', () => {
    const { asFragment } = render(<Button data-testid={buttonTestid}>Click me!</Button>)

    expect(asFragment()).toMatchSnapshot()

    const button = screen.getByTestId(buttonTestid)

    expect(button).toBeInTheDocument()
    expect(button).toHaveTextContent('Click me!')
  })

  it('Кнопка должна быть отключена, когда свойство disabled равно true', () => {
    const { asFragment } = render(
      <Button disabled data-testid={buttonTestid}>
        Click me!
      </Button>
    )

    expect(asFragment()).toMatchSnapshot()

    const button = screen.getByTestId(buttonTestid)

    expect(button).toBeInTheDocument()
    expect(button).toBeDisabled()
  })

  it('Кнопка должна быть включена, когда свойство disabled равно false', () => {
    const { asFragment } = render(
      <Button disabled={false} data-testid={buttonTestid}>
        Click me!
      </Button>
    )

    expect(asFragment()).toMatchSnapshot()

    const button = screen.getByTestId(buttonTestid)

    expect(button).toBeInTheDocument()
    expect(button).not.toBeDisabled()
  })

  it('Кнопка должна отображать иконку слева до текста, если она передана', () => {
    const icon = (
      <span role='img' aria-label='icon'>
        🌟
      </span>
    )

    const { asFragment } = render(
      <Button data-testid={buttonTestid} iconLeft={icon}>
        Click me!
      </Button>
    )

    expect(asFragment()).toMatchSnapshot()

    const button = screen.getByTestId(buttonTestid)
    const iconElement = screen.getByRole('img')
    const buttonText = screen.getByText('Click me!')

    expect(button).toBeInTheDocument()
    expect(button).toContainElement(iconElement)
    expect(button).toContainElement(buttonText)

    const buttonContent = button.textContent!
    expect(buttonContent.indexOf('🌟')).toBeLessThan(buttonContent.indexOf('Click me!'))
  })

  it('Кнопка должна отображать иконку справа после текста, если она передана', () => {
    const icon = (
      <span role='img' aria-label='icon'>
        🌟
      </span>
    )

    const { asFragment } = render(
      <Button data-testid={buttonTestid} iconRight={icon}>
        Click me!
      </Button>
    )

    expect(asFragment()).toMatchSnapshot()

    const button = screen.getByTestId(buttonTestid)
    const iconElement = screen.getByRole('img')
    const buttonText = screen.getByText('Click me!')

    expect(button).toBeInTheDocument()
    expect(button).toContainElement(iconElement)
    expect(button).toContainElement(buttonText)

    const buttonContent = button.textContent!
    expect(buttonContent.indexOf('Click me!')).toBeLessThan(buttonContent.indexOf('🌟'))
  })

  it('На кнопку добавляется новый класс, если передается className', () => {
    const { asFragment } = render(
      <Button data-testid={buttonTestid} className='testClassName'>
        Click me!
      </Button>
    )

    expect(asFragment()).toMatchSnapshot()

    const button = screen.getByTestId(buttonTestid)
    const buttonClassName = document.querySelector('.testClassName')

    expect(button).toBeInTheDocument()
    expect(buttonClassName).toBeInTheDocument()
  })

  it('Кнопка меняет тип, если передать type', () => {
    const { asFragment } = render(
      <Button data-testid={buttonTestid} type='reset'>
        Click me!
      </Button>
    )

    expect(asFragment()).toMatchSnapshot()

    const button = screen.getByTestId(buttonTestid)

    expect(button).toBeInTheDocument()
    expect(button.getAttribute('type')).toBe('reset')
  })

  it('Кнопка применяет нужные стили при передаче props textFormat', () => {
    const { asFragment } = render(
      <Button data-testid={buttonTestid} textFormat='uppercase'>
        Click me!
      </Button>
    )

    expect(asFragment()).toMatchSnapshot()

    const button = screen.getByTestId(buttonTestid)

    expect(button).toBeInTheDocument()
    expect(button).toHaveClass('uppercase')
  })
})
