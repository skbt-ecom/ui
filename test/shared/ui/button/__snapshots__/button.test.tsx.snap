// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Button > –ö–Ω–æ–ø–∫–∞ –¥–æ–ª–∂–Ω–∞ –±—ã—Ç—å –≤–∫–ª—é—á–µ–Ω–∞, –∫–æ–≥–¥–∞ —Å–≤–æ–π—Å—Ç–≤–æ disabled —Ä–∞–≤–Ω–æ false 1`] = `
<DocumentFragment>
  <button
    class="relative flex items-center justify-center gap-4 cursor-pointer rounded-sm w-[216px] outline-offset-4 outline-transparent outline-2 desk-body-regular-l transition duration-12 active:scale-[0.97] disabled:pointer-events-none bg-color-primary-default text-color-white hover:bg-color-primary-hover active:bg-color-primary-hover focus:bg-color-primary-default focus:outline-primary-focus disabled:bg-color-primary-disabled h-12 initial"
    data-testid="test-button"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Button > –ö–Ω–æ–ø–∫–∞ –¥–æ–ª–∂–Ω–∞ –±—ã—Ç—å –æ—Ç–∫–ª—é—á–µ–Ω–∞, –∫–æ–≥–¥–∞ —Å–≤–æ–π—Å—Ç–≤–æ disabled —Ä–∞–≤–Ω–æ true 1`] = `
<DocumentFragment>
  <button
    class="relative flex items-center justify-center gap-4 cursor-pointer rounded-sm w-[216px] outline-offset-4 outline-transparent outline-2 desk-body-regular-l transition duration-12 active:scale-[0.97] disabled:pointer-events-none bg-color-primary-default text-color-white hover:bg-color-primary-hover active:bg-color-primary-hover focus:bg-color-primary-default focus:outline-primary-focus disabled:bg-color-primary-disabled h-12 initial"
    data-testid="test-button"
    disabled=""
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Button > –ö–Ω–æ–ø–∫–∞ –¥–æ–ª–∂–Ω–∞ –æ—Ç–æ–±—Ä–∞–∂–∞—Ç—å –∏–∫–æ–Ω–∫—É —Å–ª–µ–≤–∞ –¥–æ —Ç–µ–∫—Å—Ç–∞, –µ—Å–ª–∏ –æ–Ω–∞ –ø–µ—Ä–µ–¥–∞–Ω–∞ 1`] = `
<DocumentFragment>
  <button
    class="relative flex items-center justify-center gap-4 cursor-pointer rounded-sm w-[216px] outline-offset-4 outline-transparent outline-2 desk-body-regular-l transition duration-12 active:scale-[0.97] disabled:pointer-events-none bg-color-primary-default text-color-white hover:bg-color-primary-hover active:bg-color-primary-hover focus:bg-color-primary-default focus:outline-primary-focus disabled:bg-color-primary-disabled h-12 initial"
    data-testid="test-button"
    type="button"
  >
    <span
      class="size-5 flex items-center justify-center"
    >
      <span
        aria-label="icon"
        role="img"
      >
        üåü
      </span>
    </span>
    Click me!
  </button>
</DocumentFragment>
`;

exports[`–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Button > –ö–Ω–æ–ø–∫–∞ –¥–æ–ª–∂–Ω–∞ –æ—Ç–æ–±—Ä–∞–∂–∞—Ç—å –∏–∫–æ–Ω–∫—É —Å–ø—Ä–∞–≤–∞ –ø–æ—Å–ª–µ —Ç–µ–∫—Å—Ç–∞, –µ—Å–ª–∏ –æ–Ω–∞ –ø–µ—Ä–µ–¥–∞–Ω–∞ 1`] = `
<DocumentFragment>
  <button
    class="relative flex items-center justify-center gap-4 cursor-pointer rounded-sm w-[216px] outline-offset-4 outline-transparent outline-2 desk-body-regular-l transition duration-12 active:scale-[0.97] disabled:pointer-events-none bg-color-primary-default text-color-white hover:bg-color-primary-hover active:bg-color-primary-hover focus:bg-color-primary-default focus:outline-primary-focus disabled:bg-color-primary-disabled h-12 initial"
    data-testid="test-button"
    type="button"
  >
    Click me!2
    <span
      class="size-5 flex items-center justify-center"
    >
      <span
        aria-label="icon"
        role="img"
      >
        üåü
      </span>
    </span>
  </button>
</DocumentFragment>
`;

exports[`–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Button > –ö–Ω–æ–ø–∫–∞ –¥–æ–ª–∂–Ω–∞ —Ä–µ–Ω–¥–µ—Ä–∏—Ç—å—Å—è —Å –ø—Ä–∞–≤–∏–ª—å–Ω—ã–º —Ç–µ–∫—Å—Ç–æ–º 1`] = `
<DocumentFragment>
  <button
    class="relative flex items-center justify-center gap-4 cursor-pointer rounded-sm w-[216px] outline-offset-4 outline-transparent outline-2 desk-body-regular-l transition duration-12 active:scale-[0.97] disabled:pointer-events-none bg-color-primary-default text-color-white hover:bg-color-primary-hover active:bg-color-primary-hover focus:bg-color-primary-default focus:outline-primary-focus disabled:bg-color-primary-disabled h-12 initial"
    data-testid="test-button"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Button > –ö–Ω–æ–ø–∫–∞ –º–µ–Ω—è–µ—Ç —Ç–∏–ø, –µ—Å–ª–∏ –ø–µ—Ä–µ–¥–∞—Ç—å type 1`] = `
<DocumentFragment>
  <button
    class="relative flex items-center justify-center gap-4 cursor-pointer rounded-sm w-[216px] outline-offset-4 outline-transparent outline-2 desk-body-regular-l transition duration-12 active:scale-[0.97] disabled:pointer-events-none bg-color-primary-default text-color-white hover:bg-color-primary-hover active:bg-color-primary-hover focus:bg-color-primary-default focus:outline-primary-focus disabled:bg-color-primary-disabled h-12 initial"
    data-testid="test-button"
    type="reset"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Button > –ö–Ω–æ–ø–∫–∞ –ø—Ä–∏–º–µ–Ω—è–µ—Ç –Ω—É–∂–Ω—ã–µ —Å—Ç–∏–ª–∏ –ø—Ä–∏ –ø–µ—Ä–µ–¥–∞—á–µ props textFormat 1`] = `
<DocumentFragment>
  <button
    class="relative flex items-center justify-center gap-4 cursor-pointer rounded-sm w-[216px] outline-offset-4 outline-transparent outline-2 desk-body-regular-l transition duration-12 active:scale-[0.97] disabled:pointer-events-none bg-color-primary-default text-color-white hover:bg-color-primary-hover active:bg-color-primary-hover focus:bg-color-primary-default focus:outline-primary-focus disabled:bg-color-primary-disabled h-12 uppercase"
    data-testid="test-button"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`–¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Button > –ù–∞ –∫–Ω–æ–ø–∫—É –¥–æ–±–∞–≤–ª—è–µ—Ç—Å—è –Ω–æ–≤—ã–π –∫–ª–∞—Å—Å, –µ—Å–ª–∏ –ø–µ—Ä–µ–¥–∞–µ—Ç—Å—è className 1`] = `
<DocumentFragment>
  <button
    class="relative flex items-center justify-center gap-4 cursor-pointer rounded-sm w-[216px] outline-offset-4 outline-transparent outline-2 desk-body-regular-l transition duration-12 active:scale-[0.97] disabled:pointer-events-none bg-color-primary-default text-color-white hover:bg-color-primary-hover active:bg-color-primary-hover focus:bg-color-primary-default focus:outline-primary-focus disabled:bg-color-primary-disabled h-12 initial testClassName"
    data-testid="test-button"
    type="button"
  >
    Click me!
  </button>
</DocumentFragment>
`;
